File: VmTranslator/src/main/java/org/example/Parser.java
package org.example;

import org.apache.commons.io.IOUtils;
import org.apache.commons.lang3.StringUtils;

import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;

public final class Parser
        implements IParser
{
    private final ArrayList<String> lines;
    private       int               currentIndex = -1;
    private       String[]          fields;

    public Parser(Path file)
    {
        lines = new ArrayList<>();
        try (InputStream is = IOUtils.toBufferedInputStream(new FileInputStream(file.toFile()))) {
            lines.addAll(IOUtils.readLines(is, StandardCharsets.UTF_8));
        } catch (IOException e) {
            System.err.println("Problem reading file " + file);
            e.printStackTrace();
            System.exit(2);
        }
    }

    @Override
    public boolean advance()
    {
        while (++currentIndex < lines.size()) {
            String nextLine = StringUtils.trim(lines.get(currentIndex));
            if (nextLine.startsWith("//") || StringUtils.isBlank(nextLine)) {
                continue;
            }
            fields = StringUtils.split(nextLine);
            return true;
        }
        return false;
    }

    @Override
    public CommandType commandType()
    {
        return switch (fields[0]) {
            case "push": yield CommandType.PUSH;
            case "pop": yield CommandType.POP;
            case "add":
            case "sub":
            case "neg":
                yield CommandType.ARITHMETIC;

        }
    }

    @Override
    public String firstArgument()
    {
        return null;
    }

    @Override
    public int secondArgument()
    {
        return 0;
    }
}

File: VmTranslator/src/main/java/ch/chassaing/hack/vm/Parser.kt
package ch.chassaing.hack.vm

import ch.chassaing.hack.vm.command.*
import org.apache.commons.lang3.StringUtils
import java.util.*

class Parser(private val lines: List<String>) : IParser {

    private var currentLine = -1
    private lateinit var fields: Array<String>

    override fun advance(): Int {
        while (++currentLine < lines.size) {
            val nextLine = StringUtils.trim(lines[currentLine])
            if (nextLine.startsWith("//") || StringUtils.isBlank(nextLine)) {
                continue
            }
            fields = StringUtils.split(nextLine)
            return currentLine
        }
        return -1
    }

    override fun command(): Command {
        val segment: Segment
        val position: Int
        return when (fields[0]) {
            "push" -> {
                segment = Segment.valueOf(fields[1].uppercase(Locale.getDefault()))
                position = fields[2].toInt()
                Push(currentLine, segment, position)
            }

            "pop" -> {
                segment = Segment.valueOf(fields[1].uppercase(Locale.getDefault()))
                position = fields[2].toInt()
                Pop(currentLine, segment, position)
            }

            "add" -> Add(currentLine)
            "sub" -> Sub(currentLine)
            "and" -> And(currentLine)
            "or" -> Or(currentLine)
            "eq" -> Eq(currentLine)
            "lt" -> Lt(currentLine)
            "gt" -> Gt(currentLine)
            "neg" -> Neg(currentLine)
            "not" -> Not(currentLine)
            else -> throw UnsupportedOperationException("Unknown command " + fields[0])
        }
    }
}